humility-cmd/src/hiffy.rs

```diff
@@ -1006,3 +998,144 @@ impl<'a> HiffyContext<'a> {
         self.scratch_size
     }
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Welcome to the Cursed RPC Zone
+
+struct HiffySendWorkspace {
+    hubris: Option<*const HubrisArchive>,
+    core: Option<*mut dyn Core>,
+
+    /// If we receive an RPC result, then record the buffer here
+    results: Vec<Vec<u8>>,
+
+    errors: Vec<anyhow::Error>,
+}
+
+// SAFETY:
+// It's safe to send this between threads, albeit unsafe to actually dereference
+// the pointers within.
+unsafe impl Send for HiffySendWorkspace {}
+
+lazy_static::lazy_static! {
+    static ref HIFFY_SEND_WORKSPACE: Arc<Mutex<HiffySendWorkspace>> = Arc::new(Mutex::new(
+        HiffySendWorkspace {
+            hubris: None,
+            core: None,
+            results: vec![],
+            errors: vec![],
+        }
+    ));
+}
+
+fn hiffy_dummy_fn(
+    _stack: &[Option<u32>],
+    _data: &[u8],
+    _rval: &mut [u8],
+) -> Result<usize, Failure> {
+    panic!("dummy function should never be called");
+}
+
+fn hiffy_send_fn(
+    stack: &[Option<u32>],
+    _data: &[u8],
+    rval: &mut [u8],
+) -> Result<usize, Failure> {
+    let mut payload = [0u8; 32];
+
+    let sp = stack.len();
+    if sp < 4 {
+        return Err(Failure::Fault(Fault::MissingParameters));
+    }
+
+    let nreply =
+        stack[sp - 1].ok_or(Failure::Fault(Fault::EmptyParameter(4)))?;
+    let nbytes: u16 = stack[sp - 2]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(3)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let fp = sp - (nbytes as usize + 4);
+
+    let task: u16 = stack[fp]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(0)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let op: u16 = stack[fp + 1]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(1)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let base = fp + 2;
+
+    for i in base..base + nbytes as usize {
+        payload[i - base] = stack[i]
+            .ok_or(Failure::Fault(Fault::EmptyParameter(2)))?
+            .try_into()
+            .map_err(|_| Failure::Fault(Fault::BadParameter(2)))?;
+    }
+
+    let (hubris, core) = {
+        let workspace = HIFFY_SEND_WORKSPACE.lock().unwrap();
+        // SAFETY: we only ever call this function when the pointers are
+        // populated, and reset them to None afterwards.  This means we should
+        // fail at the initial unwrap() if someone violates the rules.
+        unsafe {
+            (
+                workspace.hubris.unwrap().as_ref().unwrap(),
+                workspace.core.unwrap().as_mut().unwrap(),
+            )
+        }
+    };
```

Comment:
Unlocking `HIFFY_SEND_WORKSPACE` while we hold references to data it contains seems really dicey, even given the warnings about the cursedness of this code. :) Could we lock `HIFFY_SEND_WORKSPACE` once and keep it locked for the duration of this function instead? At least then we know we're not going to accidentally create multiple references to the archive or core.

Reply:
ah yeah, that's a good catch – I was aggressively unlocking it because otherwise a panic in this function causes mutex poisoning, but that seems sketchy.  Take a look at f23209896c885c36fecf2ecdc86567da96b469b3

Reply:
That looks good, thanks!